### PR TITLE
fix: Deduplicate assembly references by name and add architecture overview to CLAUDE.md

### PR DESCRIPTION
**Pull request by hatayama**
## Summary

- Add Architecture Overview section to CLAUDE.md to clarify that this project is a CLI tool communicating via TCP, not an MCP tool. The namespace `uLoopMCP` is historical — AI agents previously misidentified the project as MCP-based due to the naming.
- Change assembly reference deduplication from file-path-based to assembly-name-based in `AssemblyBuilderCompiler`. The same assembly (e.g. `System.Threading`) can exist under both MonoBleedingEdge and NetStandard directories with different paths, which could cause duplicate identity compilation errors during dynamic code execution.

### CLAUDE.md (AGENTS.md)
- Move Architecture Overview to the top of the file
- Explain the CLI-based architecture and historical namespace naming

### AssemblyBuilderCompiler.cs
- `GetCachedAppDomainReferences`: use `Assembly.GetName().Name` as dedup key instead of file path
- `CollectReferences`: extract `MergeReferencesByAssemblyName` as `internal static` method using `Path.GetFileNameWithoutExtension` as dedup key
- Remove redundant null check for `additionalRefs` (already guarded by early return)

### AssemblyBuilderCompilerMergeReferencesTests.cs (new)
- 6 unit tests covering: same-name dedup across base/additional refs, base-only dedup, case insensitivity, non-existent file skipping, null/empty path handling

## Test plan
- [x] `uloop compile` passes with 0 errors and 0 warnings
- [x] All 6 new tests pass (`uloop run-tests --filter-type regex --filter-value AssemblyBuilderCompilerMergeReferencesTests`)
- [x] Red phase confirmed: reverting to path-based dedup causes 3 of 6 tests to fail

---
## Summary by cubic
Deduplicates assembly references by assembly name to prevent CS1703 duplicate-identity errors during dynamic compilation. Adds an Architecture Overview to `AGENTS.md` clarifying this is a CLI tool (not MCP) and the historical `io.github.hatayama.uLoopMCP` namespace.

- **Bug Fixes**
  - `AssemblyBuilderCompiler`: dedupe by assembly name in AppDomain scan and during merge; introduce internal static `MergeReferencesByAssemblyName`.
  - Case-insensitive matching; skip null/empty paths and non-existent files.
  - Add 6 unit tests for name-based dedup, case-insensitivity, and path validation.
  - Add `Debug.Assert` preconditions in `MergeReferencesByAssemblyName`; strengthen null/empty path test; remove an unused using.

<sup>Written for commit c61b798b755c9b74b9509c19589b422cc6f84345. Summary will update on new commits.</sup>

## Overview

This PR switches assembly reference deduplication from file-path-based to assembly-name-based and adds an architecture overview to AGENTS.md clarifying the project is a CLI tool.

### 1. Assembly Reference Deduplication
- Problem: Path-based deduplication allowed the same assembly to be referenced multiple times from different locations (e.g., MonoBleedingEdge vs NetStandard), causing CS1703 duplicate-identity compilation errors.
- Solution:
  - AssemblyBuilderCompiler.GetCachedAppDomainReferences now deduplicates cached assemblies by Assembly.GetName().Name instead of asm.Location.
  - Added internal static MergeReferencesByAssemblyName(string[] baseReferences, List<string> additionalRefs) to AssemblyBuilderCompiler. It:
    - Uses Path.GetFileNameWithoutExtension(path) as the dedup key (case-insensitive, StringComparer.OrdinalIgnoreCase).
    - Preserves the first occurrence from baseReferences.
    - Adds additionalRefs only if the file path exists and is non-empty, skipping null/empty/whitespace entries.
  - CollectReferences() delegates to MergeReferencesByAssemblyName and a redundant null-check for additionalRefs was removed.

### 2. Documentation
- AGENTS.md: Added an "Architecture Overview" section at the top:
  - States the project exposes a CLI tool (uloop) that communicates with Unity Editor via TCP.
  - Instructs AI agents to interact via CLI commands (e.g., uloop get-logs, uloop compile) rather than MCP.
  - Notes the TypeScript MCP server exists but is not the primary interface.
  - Clarifies the C# namespace remains io.github.hatayama.uLoopMCP for historical reasons despite being CLI-based.

## Tests

Added NUnit fixture AssemblyBuilderCompilerMergeReferencesTests (Assets/Tests/Editor/DynamicCodeToolTests/AssemblyBuilderCompilerMergeReferencesTests.cs) with 6 tests:
1. Keep first occurrence when same assembly name exists in base and additional refs.
2. Include both when assembly names differ.
3. Deduplicate duplicate assembly names within base references.
4. Case-insensitive assembly-name comparison.
5. Skip additional references that do not exist on disk.
6. Skip additional references that are null/empty/whitespace.

All tests exercise MergeReferencesByAssemblyName behavior.

## Test Results / Notes
- New tests cover the deduplication behavior and edge cases (non-existent and empty paths).
- Changes prevent CS1703 by deduplicating by assembly name instead of path.

## Files Changed (high level)
- AGENTS.md: +8/-2 (added Architecture Overview)
- Packages/src/Editor/Compilation/AssemblyBuilderCompiler.cs: added MergeReferencesByAssemblyName and updated reference deduplication logic
- Assets/Tests/Editor/DynamicCodeToolTests/AssemblyBuilderCompilerMergeReferencesTests.cs: +134 (new tests)

### CODE REVIEW

File: AGENTS.md [hunk 1]
```diff
@@ -1,6 +1,12 @@
-Comments in the code, commit messages, and PR titles and bodies should be written in English.
+## Architecture Overview
+
+This project provides a **CLI tool (`uloop`)** that communicates with Unity Editor via TCP.
+AI agents interact with Unity through `uloop` CLI commands (e.g., `uloop get-logs`, `uloop compile`), NOT through MCP protocol directly.
+The MCP server (TypeScriptServer~) exists as a separate component but is not the primary interface for AI agents.
```
[the remark this answers is not in the document]
Comment (hatayama) [reply AGENTS.md:5]: The tilde is intentional — `TypeScriptServer~` is the actual directory name in the project. Unity uses the `~` suffix convention to exclude folders from compilation. No change needed.

File: Packages/src/Editor/Compilation/AssemblyBuilderCompiler.cs [hunk 2]
```diff
@@ -274,25 +274,29 @@ private string[] CollectReferences(List<string> additionalRefs)
...
+                    refs.Add(baseRef);
                 }
             }
 
-            if (additionalRefs != null)
+            foreach (string refPath in additionalRefs)
             {
-                foreach (string refPath in additionalRefs)
+                string name = Path.GetFileNameWithoutExtension(refPath);
+                if (!string.IsNullOrEmpty(refPath) && File.Exists(refPath) && seenNames.Add(name))
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Packages/src/Editor/Compilation/AssemblyBuilderCompiler.cs]: Already addressed in 83cb87e (guard moved before `Path.GetFileNameWithoutExtension`). Additionally, added `Debug.Assert` preconditions in c61b798b.
